### ISSUE TITLE
(Chrome) Extension dies on extension reloads

### ISSUE THREAD

**Issue by Aquafina-water-bottle** [themoeway/yomitan#114]
As mentioned in #112, extensions are no longer persistent, and can be reloaded at any time. Unfortunately, on Chromium, if Yomitan does refresh, then the extension simply fails to work until re-opening the browser.

The following instructions are how to force a reload of the extension on Chromium.

> Taken from: https://stackoverflow.com/a/45947062

1. Add a print message somewhere in the source code for easier debugging, i.e. under `ext/js/background/backend.js`:

    ```js
    async _openWelcomeGuidePage() {
        console.log('Hello world', new Date()); // <-- add this line!
        await this._createTab(chrome.runtime.getURL('/welcome.html'));
    }
    ```

2. Download the add-on from folder (`ext`) from a clean state (delete the Yomitan addon if it was already installed, and temporarily disable Yomichan). Ensure that the welcome screen shows on startup (or on clicking the popup button).
3. Cloze Chromium and re-open. Usually, a welcome screen should show. If not, head over to the Yomitan settings.
4. Edit any file in the extension, and save. For example, change the `console.log` message we just added:

    ```js
    async _openWelcomeGuidePage() {
        console.log('A different message', new Date());
        await this._createTab(chrome.runtime.getURL('/welcome.html'));
    }
    ```

5. Back in Chromium, right click anywhere, and then inspect the page. Having the dev tools open is required to do the step below.
6. Right click the refresh button, and click `Empty Cache and Hard Reset`
7. Wait for a few seconds, and then navigate to the Yomitan settings. Another welcome screen should show (assuming #113 is not merged).

By this step, the extension completely breaks. Scanning doesn't work, settings cannot be loaded, the popup buttons are unresponsive, etc. It seems to be silently failing as well, because no errors are shown on the console.

---

Note: Although I couldn't figure out how to force a reload on Firefox instantly, you can force it to reload by idling for about 5 minutes. Fortunately, nothing seemed to break on Firefox after one reload. In other words, this appears to a Chromium only issue.

**Comment by djahandarie** [comment]
Maybe the best short-term approach would be to hack around it using a workaround like the `offscreen` trick mentioned here: https://stackoverflow.com/a/66618269

Making yomitan entirely resilient to reloads might be quite complicated, though if we can make some progress on understanding the size of work that'd be required, that would also be good. I wonder if it's even possible though, because it needs to be constantly monitoring all behavior on webpages, so it does essentially need to be always on...

**Comment by donkuri** [comment]
Tested on the latest version of Chromium on Arch Linux (EndeavourOS) to be exact, the same exact bug occurs: The controls become unresponsive and it tells me that the page is taking too long to load.

**Comment by Aquafina-water-bottle** [comment]
Quick note on this: I don't think the constantly pinging hack would work, since at least on my end, Chrome sometimes refreshes the add-on nearly instantly when opening the first instance of Chrome (and essentially killing the add-on immediately). I think the expected behavior should be it only reloads after 30s minimum, but Chrome seems to like doing its own thing :shrug: 
The only hack that I think could work would be the much more intrusive "open a visible tab" approach, and have that replace background.html.

**Comment by Aquafina-water-bottle** [comment]
Potentially related: On Chromium, the preview popup within the settings never properly refreshes once the service worker stops. Untested on Firefox.

**Comment by praschke** [comment]
This issue is confusing two different things. The process described is an unpacked extension update via hard reload, which causes a second service worker to be installed because the sources differ.

### Update

When the new service worker is installed, a number of weird behaviors can happen. The old worker can be kept alive and the new worker will be kept in limbo. Keeping devtools open can apparently cause this, but keeping an extension page open here seems to keep the old worker alive without devtools, sometimes even when stopped.

At some point the old worker will be killed and the new worker might be started, working normally, or it might stay stopped. If it stays stopped everything stays broken, extension pages won't be opened from the action popup, the toggle might also refuse to be flipped but that's inconsistent. Starting the new worker manually in devtools seems to take care of the broken behavior and everything starts normally upon reloading, though this is also somewhat inconsistent.

Restarting the browser does not guarantee Yomitan is started with the fresh sources. I've noticed the extension can be updated shortly after starting the browser, after the old service worker has already established itself.

I don't expect this behavior to show up to end users, who typically aren't using/modifying unpacked extensions. All of this stale service worker nonsense also seems to be mitigated by instead pressing the reload button in the chrome extensions page, which closes open extension pages and kills the old context right away.

### Stopping/Idling

Updating is not the same as the service worker routinely stopping and starting, which you can trigger manually in the application tab of devtools. The behavior here seems considerably more benign.

When the browser routinely stops the worker, opening the action popup and extension pages starts it, scanning a page starts it, navigating starts it. Changing options on all pages starts it. Entering text on the search page starts it. When the search page is open and clipboard scanning is on, the worker will never stop.

The popup preview in the settings page does not respond after the service worker has been stopped once. The settings page has to be reloaded for the preview to work again.

Altogether, this is not such a showstopper.

**Comment by praschke** [comment]
the process described no longer replicates on version 117. on `Empty Cache and Hard Reload` the new service worker is not installed and the old service worker remains active and functional. using the reload button on the chrome extensions page seems to only install the new service worker onde all clients (extension pages and content scripts) have been removed.

this seems to have been a Chrome bug and has now been fixed.

**Comment by djahandarie** [comment]
Given that the chrome bug is fixed, and we also have fixed #241 (the inactive service worker issue), I will close this too.

Thanks all for your help closing this set of high priority issues :pray: we are very close to a stable release now!
